### PR TITLE
DOC: run typing checks

### LINKED ISSUE #51313: Instructions for checking typing of code cause pylint and autotyping to run
At https://pandas.pydata.org/docs/dev/development/contributing_codebase.html#validating-type-hints we tell people to validate the types of their code by doing:  `pre-commit run --hook-stage manual --all-files` .  The problem here is that this causes `autotyping` and `pylint` to run.

`pylint` takes a long time, so we should tell people to skip it.
Worse is that `autotyping` causes lots of files to change, so you end up with a bigger set of changed files.

I think we should modify the `.pre-commit-config.yaml` to define a `typing` stage that only runs `pyright` and `mypy`.

Even if we do that, I get a lot of typing errors on the code in `main` from both type checkers.  So that should be addressed as well.

### PR DESCRIPTION
**Pull request by twoertwein**
- [ ] closes #51313 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: doc/source/development/contributing_codebase.rst [hunk 1]
```diff
@@ -270,13 +270,17 @@ pandas uses `mypy <http://mypy-lang.org>`_ and `pyright <https://github.com/micr
...
+    pre-commit run --hook-stage manual --all-files stubtest
 
-    # if the above fails due to stubtest
-    SKIP=stubtest pre-commit run --hook-stage manual --all-files
+in your active python environment.
 
-in your activated python environment. A recent version of ``numpy`` (>=1.22.0) is required for type validation.
+.. warning::
+
+    * It is essential that the versions of ``mypy`` and ``numpy`` in the active environment are the same as in ``requirements-dev.txt``.
```
Comment (Dr-Irv) [inline doc/source/development/contributing_codebase.rst:283]: There is no version number of `numpy` in `requirements-dev.txt` or `environment.yml` . 

I think we should be adding a statement about updating the environment.   For conda environments, you need to de-activate the environment, and do `mamba env update -f environment.yml` .  Not sure how to do the updates when using `pip`, so we should indicate both ways.  See my comment here:  https://github.com/pandas-dev/pandas/issues/51313#issuecomment-1427056543

**Review by Dr-Irv — CHANGES_REQUESTED** [review]
In line 269, we have this:
"after making any change you can ensure your type hints are correct by running"

What I've found is that you have to stage your changes via `git add` in order for the commands you provided to work.  So I think the docs should be modified to say that as well.

File: doc/source/development/contributing_codebase.rst (hunk 1 above)
Comment (twoertwein) [reply doc/source/development/contributing_codebase.rst]: I added a link to the python setup docs.

> What I've found is that you have to stage your changes via git add in order for the commands you provided to work. 

When adding  `--all-files` this does not seem to be needed. Without it pre-commit says:

> [WARNING] Unstaged files detected.
> [INFO] Stashing unstaged files to /var/home/twoertwein/.cache/pre-commit/patch1676322301-669741.

Comment (Dr-Irv) [reply doc/source/development/contributing_codebase.rst]: > When adding `--all-files` this does not seem to be needed. Without it pre-commit says:

Hmmm.  Maybe I forgot to include that.

File: doc/source/development/contributing_codebase.rst [hunk 2]
```diff
@@ -270,13 +270,17 @@ pandas uses `mypy <http://mypy-lang.org>`_ and `pyright <https://github.com/micr
...
+    pre-commit run --hook-stage manual --all-files stubtest
 
-    # if the above fails due to stubtest
-    SKIP=stubtest pre-commit run --hook-stage manual --all-files
+in your active python environment.
 
-in your activated python environment. A recent version of ``numpy`` (>=1.22.0) is required for type validation.
+.. warning::
+
+    * It is essential that the versions of ``mypy`` and ``numpy`` are the same as used on main. Please see :ref:`how to setup the python environment <contributing.mamba>`.
```
[the remark this answers is not in the document]
Comment (Dr-Irv) [reply doc/source/development/contributing_codebase.rst:283]: > I don't use mamba/conda so I don't feel comfortable commenting on how to update the environment.
> 
OK, maybe I will do a PR about that.

> My simple/stupid approach: If mypy/pyright fails, I just check which numpy/mypy version main installs for the most recently merged PR and then pip install the same version :) Just matching the numpy/mypy version has so far always worked for me.

Yes, but you're smarter about these things than our average contributor!  I've found these kinds of issues a PITA to figure out what is wrong with my environment.

[the remark this answers is not in the document]
Comment (Dr-Irv) [reply doc/source/development/contributing_codebase.rst:283]: I still think we should delete `numpy` from this part of the doc.

[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply doc/source/development/contributing_codebase.rst:283]: Hi @Dr-Irv - what do you suggest this be replaced with if you don't want to mention numpy?

I also use the approach of checking which version was running in CI, and checking my environment matches that. I think that'd be fine to document

[the remark this answers is not in the document]
Comment (Dr-Irv) [reply doc/source/development/contributing_codebase.rst:283]: > Hi @Dr-Irv - what do you suggest this be replaced with if you don't want to mention numpy?
> 

My concern here is that the sentence is saying "what is used in main", but there is no specific version on `main` .  The version is determined by what is used in CI.  So I'm not exactly sure what to say.

> I also use the approach of checking which version was running in CI, and checking my environment matches that. I think that'd be fine to document

Think of it from a first-time contributor point of view.  How are they supposed to figure out what was running in CI when they haven't pushed anything yet?  Documenting how to look in the CI (or somewhere else?) would be fine to add here.

[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply doc/source/development/contributing_codebase.rst:283]: Agreed, rephrasing "what is used in main" to something to the effect of "what was used in the latest CI build", and mentioning which CI job they should be looking at (i.e. the one which runs mypy)

[the remark this answers is not in the document]
Comment (twoertwein) [reply doc/source/development/contributing_codebase.rst]: I tried to describe how to navigate to the section of a pandas CI report where the versions are listed (not sure whether that is clear enough - probably not).

File: doc/source/development/contributing_codebase.rst [hunk 3]
```diff
@@ -266,17 +266,21 @@ This module will ultimately house types for repeatedly used concepts like "path-
...
+    pre-commit run --hook-stage manual --all-files stubtest
 
-    # if the above fails due to stubtest
-    SKIP=stubtest pre-commit run --hook-stage manual --all-files
+in your python environment.
 
-in your activated python environment. A recent version of ``numpy`` (>=1.22.0) is required for type validation.
+.. warning::
+
+    * Please be aware that the above commands will use the current python environment. If your python packages are older/newer than those installed by the pandas CI, the above commands might fail. This is often the case when the ``mypy`` or ``numpy`` versions do not match. Please see :ref:`how to setup the python environment <contributing.mamba>` or click on a recent CI report, then click on "Set up Conda" and "Environment info" to see which versions the pandas CI installs.
```
Comment (Dr-Irv) [inline doc/source/development/contributing_codebase.rst:283]: Might want to say "Look at https://github.com/pandas-dev/pandas/actions/workflows/code-checks.yml for a successful code check, then click on  'Docstring validation, typing, and other manual pre-commit hooks' , then click on Set up Conda" and "Environment info" to see which versions the pandas CI installs."

Comment (twoertwein) [reply doc/source/development/contributing_codebase.rst]: Thank you!

**Review by Dr-Irv — APPROVED** [review]
Thanks @twoertwein

**Pull request merged**